tsdb/agent/db_append_v2.go

```diff
@@ -0,0 +1,224 @@
+// Copyright The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package agent
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/prometheus/prometheus/model/exemplar"
+	"github.com/prometheus/prometheus/model/histogram"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/value"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/tsdb"
+	"github.com/prometheus/prometheus/tsdb/chunks"
+	"github.com/prometheus/prometheus/tsdb/record"
+)
+
+// AppenderV2 implements storage.AppenderV2.
+func (db *DB) AppenderV2(context.Context) storage.AppenderV2 {
+	return db.appenderV2Pool.Get().(storage.AppenderV2)
+}
+
+type appenderV2 struct {
+	appenderBase
+}
+
+// Append appends pending sample to agent's DB.
+// TODO: Wire metadata in the Agent's appender.
+func (a *appenderV2) Append(ref storage.SeriesRef, l labels.Labels, st, t int64, v float64, h *histogram.Histogram, fh *histogram.FloatHistogram, opts storage.AOptions) (storage.SeriesRef, error) {
+	var (
+		// Avoid shadowing err variables for reliability.
+		valErr, partialErr error
+		sampleMetricType   = sampleMetricTypeFloat
+		isStale            bool
+	)
+	// Fail fast on incorrect histograms.
+	switch {
+	case fh != nil:
+		sampleMetricType = sampleMetricTypeHistogram
+		valErr = fh.Validate()
+	case h != nil:
+		sampleMetricType = sampleMetricTypeHistogram
+		valErr = h.Validate()
+	}
+	if valErr != nil {
+		return 0, valErr
+	}
+
+	// series references and chunk references are identical for agent mode.
+	s := a.series.GetByID(chunks.HeadSeriesRef(ref))
+	if s == nil {
+		// Ensure no empty or duplicate labels have gotten through. This mirrors the
+		// equivalent validation code in the TSDB's headAppender.
+		l = l.WithoutEmpty()
+		if l.IsEmpty() {
+			return 0, fmt.Errorf("empty labelset: %w", tsdb.ErrInvalidSample)
+		}
+
+		if lbl, dup := l.HasDuplicateLabelNames(); dup {
+			return 0, fmt.Errorf(`label name "%s" is not unique: %w`, lbl, tsdb.ErrInvalidSample)
+		}
+
+		var created bool
+		s, created = a.getOrCreate(l)
+		if created {
+			a.pendingSeries = append(a.pendingSeries, record.RefSeries{
+				Ref:    s.ref,
+				Labels: l,
+			})
+
+			a.metrics.numActiveSeries.Inc()
+		}
+	}
+
+	s.Lock()
+	lastTS := s.lastTs
+	s.Unlock()
```

Comment:
Do you need to add a defer or to copy off some more data here? `bestEffortAppendSTZeroSample` takes the memSeries and will reference the ref all the time and the lset if there's an error.